kotlinx-coroutines-test/common/src/TestBuilders.kt

```diff
@@ -228,7 +229,26 @@ internal suspend fun <T: AbstractCoroutine<Unit>> runTestCoroutine(
                     // we expect these and will instead throw a more informative exception just below.
                     emptyList()
                 }.throwAll()
-                throw UncompletedCoroutinesError("The test coroutine was not completed after waiting for $dispatchTimeoutMs ms")
+                var completing: Boolean
+                val completionCause = try {
+                    coroutine.tryGetCompletionCause().also { completing = true }
+                } catch (e: Throwable) {
+                    completing = false
+                    null
+                }
+                var message = "After waiting for $dispatchTimeoutMs ms"
```

Comment:
It would be nice to follow our overall approach and extract slow-path to a separate method

Reply:
Agree, then maybe it's worth extracting the whole onTimeout body?